pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -155,6 +156,49 @@ func TestPriorityQueue_AddWithReversePriorityLessFunc(t *testing.T) {
 	}
 }
 
+func popPod(t *testing.T, q *PriorityQueue, pod *v1.Pod) *framework.QueuedPodInfo {
+	p, err := q.Pop()
+	if err != nil {
+		t.Fatalf("Pop failed: %v", err)
+	}
+	if p.Pod.UID != pod.UID {
+		t.Errorf("Unexpected popped pod: %v", p)
+	}
+	return p
+}
+
+func TestPop(t *testing.T) {
+	pod := st.MakePod().Name("targetpod").UID("pod1").Obj()
+	plugin := "fooPlugin1"
+	m := map[framework.ClusterEvent]sets.String{PvAdd: sets.NewString(plugin)}
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	q := NewTestQueueWithObjects(ctx, newDefaultQueueSort(), []runtime.Object{pod}, WithClusterEventMap(m))
+
+	// Disable backoff. Otherwise MoveAllToActiveOrBackoffQueue below will pick the backoff queue
+	// when running this test quickly.
+	q.podInitialBackoffDuration = 0
+
+	q.Add(pod)
+
+	// Simulate failed attempt that makes the pod unschedulable.
+	poppedPod := popPod(t, q, pod)
+	poppedPod.UnschedulablePlugins = sets.NewString(plugin)
+	if err := q.AddUnschedulableIfNotPresent(poppedPod, q.SchedulingCycle()); err != nil {
+		t.Errorf("Unexpected error from AddUnschedulableIfNotPresent: %v", err)
+	}
+
+	// Activate it again.
+	q.MoveAllToActiveOrBackoffQueue(PvAdd, nil)
+
+	// Now check result of Pop.
+	poppedPod = popPod(t, q, pod)
+	if len(poppedPod.UnschedulablePlugins) > 0 {
```

Comment:
Any chance that this is flaky? In the sense that the pod would first go to backoff and a separate go-routine moves it back to active?

Reply:
Are there any additional goroutines? This just instantiates the queue and calls some of its API methods. There are no concurrent informers or anything like that.

I think this fully deterministic.

Reply:
I was mostly wondering if setting podInitialBackoffDuration to zero is enough to ensure the Pod never makes it to Backoff, even if momentarily.

Reply:
It looks like that won't happen

https://github.com/kubernetes/kubernetes/blob/a72cbbe160b0cc8d3bceb401487c4639d42f7f4f/pkg/scheduler/internal/queue/scheduling_queue.go#L716-L732